src/datatablerow/datatablerow.jl

```diff
@@ -60,41 +64,41 @@ Base.hash(r::DataTableRow, h::UInt = zero(UInt)) = rowhash(r.dt, r.row, h)
 # comparison of DataTable rows
 # only the rows of the same DataTable could be compared
 # rows are equal if they have the same values (while the row indices could differ)
+# returns Nullable{Bool}
+# if all non-null values are equal, but there are nulls, returns null
 @compat(Base.:(==))(r1::DataTableRow, r2::DataTableRow) = isequal(r1, r2)
 
 function Base.isequal(r1::DataTableRow, r2::DataTableRow)
-    r1.dt == r2.dt || throw(ArgumentError("Comparing rows from different frames not supported"))
-    r1.row == r2.row && return true
-    for col in columns(r1.dt)
-        if !isequal(col[r1.row], col[r2.row])
-            return false
-        end
-    end
-    return true
+    isequal_row(r1.dt, r1.row, r2.dt, r2.row)
 end
 
 # internal method for comparing the elements of the same data table column
 isequal_colel(col::AbstractArray, r1::Int, r2::Int) =
     (r1 == r2) || isequal(Base.unsafe_getindex(col, r1), Base.unsafe_getindex(col, r2))
 
-function isequal_colel{T}(col::Union{NullableArray{T}, AbstractNullableCategoricalArray{T}}, r1::Int, r2::Int)
-    (r1 == r2) && return true
-    isequal(col[r1], col[r2])
-end
-
 isequal_colel(a::Any, b::Any) = isequal(a, b)
 isequal_colel(a::Nullable, b::Any) = !isnull(a) & isequal(unsafe_get(a), b)
 isequal_colel(a::Any, b::Nullable) = isequal_colel(b, a)
 isequal_colel(a::Nullable, b::Nullable) = isequal(a, b)
 
 function isequal_row(dt1::AbstractDataTable, r1::Int, dt2::AbstractDataTable, r2::Int)
-    (dt1 === dt2) && return isequal(dt1[r1], dt1[r2])
-    (ncol(dt1) == ncol(dt2)) ||
+    if dt1 === dt2
+        if r1 == r2
+            return true
+        else
+            @inbounds for (col1, col2) in zip(columns(dt1), columns(dt2))
```

Comment:
You can avoid writing that loop twice by putting it outside of the `dt1 === dt2` and `ncol(dt1) == ncol(dt2)` checks: they would just `return true` (for the former) or throw an error (for the latter), and if no then the general loop would apply.